engines/ic3sa.cpp

```diff
@@ -0,0 +1,880 @@
+/*********************                                                  */
+/*! \file ic3sa.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief IC3 with Syntax-Guided Abstraction based on
+**
+**        Model Checking of Verilog RTL Using IC3 with Syntax-Guided
+**        Abstraction.
+**            -- Aman Goel, Karem Sakallah
+**
+**
+**  within Pono, we are building on the bit-level IC3 instead of directly
+**  on IC3Base, because a lot of the functionality is the same
+**  In particular, we don't need to override inductive generalization
+**
+**/
+
+#include "engines/ic3sa.h"
+
+#include "assert.h"
+#include "core/rts.h"
+#include "smt-switch/utils.h"
+#include "utils/logger.h"
+#include "utils/term_analysis.h"
+#include "utils/term_walkers.h"
+
+using namespace smt;
+using namespace std;
+
+// TODO add option to use interpolants in inductive generalization
+
+namespace pono {
+
+unordered_set<PrimOp> controllable_ops(
+    { And,
+      Or,
+      Implies,
+      // include bit-vector versions for boolector
+      // will prune out based on sort if
+      // not-applicable e.g. for other solvers
+      BVOr,
+      BVAnd });
+
+// main IC3SA implementation
+
+IC3SA::IC3SA(const Property & p,
+             const TransitionSystem & ts,
+             const smt::SmtSolver & solver,
+             PonoOptions opt)
+    : super(p, RelationalTransitionSystem(solver), solver, opt),
+      conc_ts_(ts, to_prover_solver_),
+      f_unroller_(conc_ts_, 0),  // zero means pure-functional unrolling
+      boolsort_(solver_->make_sort(BOOL))
+{
+  engine_ = Engine::IC3SA_ENGINE;
+}
+
+IC3Formula IC3SA::get_model_ic3formula() const
+{
+  UnorderedTermSet cube_lits;
+  // first populate with predicates
+  for (const auto & p : predset_) {
+    if (solver_->get_value(p) == solver_true_) {
+      cube_lits.insert(p);
+    } else {
+      cube_lits.insert(solver_->make_term(Not, p));
+    }
+  }
+
+  EquivalenceClasses ec = get_equivalence_classes_from_model(ts_.statevars());
```

Comment:
we are relying on the compiler that it doesn't create a new object and copies the elements. We should make sure that it is the case.

Reply:
Gotcha, like pass it as a set instead?

Reply:
yeah

Reply:
That didn't make sense -- I meant pass it as a reference to get updated?

Reply:
pass it as a reference and update in the method

Reply:
Done with 6296864291b91773f118cfc5f19574921297eceb.